### PR TITLE
store/tikv: refine backoff log

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

current, backoff's debug log doesn't contain `start_ts` and `backtype` info, so it's hard to find out which stmt effect by backoff operation.

this PR add `backoff-type` and `start_ts` for `CopClient`, `Snapshot`, `Scanner` code path's backoff debug log.

so we can got sql from other stmt via `start_ts`, and know the retry reason by `type`.

### What is changed and how it works?

- annotate `start_ts` in `context.Context`
- output backoff type string

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test

Code changes

 - Add log

Side effects

 - N/A

Related changes

 - Need to cherry-pick to the release branch

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8207)

### CODE REVIEW

File: store/tikv/backoff.go [hunk 1]
```diff
@@ -228,7 +231,11 @@ func (b *Backoffer) Backoff(typ backoffType, err error) error {
 	b.totalSleep += f(b.ctx)
 	b.types = append(b.types, typ)
 
-	log.Debugf("%v, retry later(totalsleep %dms, maxsleep %dms)", err, b.totalSleep, b.maxSleep)
+	var startTs interface{} = ""
+	if ts := b.ctx.Value(TxnStartKey); ts != nil {
+		startTs = ts
+	}
+	log.Debugf("%v, retry later(totalsleep %dms, maxsleep %dms), type: %s, start_ts: %v", err, b.totalSleep, b.maxSleep, typ.String(), startTs)
```
Comment (zz-jason) [inline store/tikv/backoff.go]: s/start_ts/txn_start_ts/

Comment (zz-jason) [reply store/tikv/backoff.go]:
```
[jianzhang.zj:~/opt/gopath/src/github.com/pingcap/tidb] git:(master ✔)
➜ grepgo "\<start_ts\>"
./store/tikv/2pc.go:605:                err = errors.Errorf("con:%d Invalid transaction tso with start_ts=%v while commit_ts=%v",
./session/session.go:1445:              log.Infof("[GENERAL_LOG] con:%d user:%s schema_ver:%d start_ts:%d sql:%s%s",

[jianzhang.zj:~/opt/gopath/src/github.com/pingcap/tidb] git:(master ✔)
➜ grepgo "\<txn_start_ts\>"
./executor/adapter.go:370:                      "[QUERY] %vcost_time:%v %s succ:%v con:%v user:%s txn_start_ts:%v database:%v %v%vsql:%v",
./executor/adapter.go:374:                      "[SLOW_QUERY] %vcost_time:%v %s succ:%v con:%v user:%s txn_start_ts:%v database:%v %v%vsql:%v",
./store/tikv/coprocessor.go:633:        logStr := fmt.Sprintf("[TIME_COP_PROCESS] resp_time:%s txn_start_ts:%d region_id:%d store_addr:%s", costTime, worker.req.StartTs, task.region.id, task.storeAddr)
./store/tikv/coprocessor.go:733:                log.Warnf("txn_start_ts:%d region_id:%d store_addr:%s, coprocessor err: %v", worker.req.StartTs, task.region.id, task.storeAddr, err)
```

File: store/tikv/backoff.go [hunk 2]
```diff
@@ -185,6 +185,9 @@ type Backoffer struct {
 	vars       *kv.Variables
 }
 
+// TxnStartKey is a key for transaction start_ts info in context.Context.
+const TxnStartKey = "_txn_start_key"
```
Comment (zz-jason) [inline store/tikv/backoff.go]: no need to export

**Comment by lysu** [comment]
@zz-jason PTAL thx~

**Pull request merged**